target/snitch_cluster/sw/snax/gemm/src/snax-gemm-lib.c

```diff
@@ -0,0 +1,210 @@
+// Copyright 2023 KU Leuven.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+//
+// Xiaoling Yi <xiaoling.yi@esat.kuleuven.be>
+
+#include "snax-gemm-lib.h"
+
+#include "snax-gemm-params.h"
+
+int32_t gen_size_config(uint8_t Batch, uint8_t M, uint8_t K, uint8_t N) {
+    return ((int32_t)Batch << 24) | ((int32_t)M << 16) | ((int32_t)K << 8) |
+           (int32_t)N;
+}
+
+void base_gemm(uint8_t m, uint8_t k, uint8_t n, int8_t* A, int8_t* B,
+               int32_t* C_cpu, bool clear) {
+    for (int i = 0; i < m; i++) {
+        for (int j = 0; j < n; j++) {
+            // clear memory first before start matrix multiplication
+            // to accumulate in K dimension
+            if (clear == true) {
+                C_cpu[i * n + j] = 0;
+            }
+            for (int s = 0; s < k; s++) {
+                C_cpu[i * n + j] = C_cpu[i * n + j] + (int32_t)A[i * k + s] *
+                                                          (int32_t)B[s + j * k];
+            }
+        }
+    }
+};
+
+void batch_gemm_cpu(uint8_t Batch, uint8_t M, uint8_t K, uint8_t N, int8_t* A,
+                    int8_t* B, int32_t* C, uint32_t strideInnermostA,
+                    uint32_t strideInnermostB, uint32_t strideInnermostC,
+                    uint32_t ldA, uint32_t ldB, uint32_t ldC, uint32_t strideA,
+                    uint32_t strideB, uint32_t strideC) {
+    int8_t* start_addr_a = A;
+    int8_t* start_addr_b = B;
+    int32_t* start_addr_c = C;
+    int8_t* addr_a;
+    int8_t* addr_b;
+    int32_t* addr_c;
+
+    bool clear;
+
+    for (int b = 0; b < Batch; b++) {
+        for (int m = 0; m < M; m++) {
+            for (int n = 0; n < N; n++) {
+                for (int k = 0; k < K; k++) {
+                    // generate the start address of each sub-matrix
+                    // according to the strides definition
+                    addr_a = start_addr_a +
+                             (b * strideA + m * ldA + k * strideInnermostA) /
+                                 sizeof(int8_t);
+                    addr_b = start_addr_b +
+                             (b * strideB + n * ldB + k * strideInnermostB) /
+                                 sizeof(int8_t);
+                    addr_c = start_addr_c +
+                             (b * strideC + m * ldC + n * strideInnermostC) /
+                                 sizeof(int32_t);
```

Comment:
What happens if this is not divisible by 4?

Reply:
In normal, this should be divisible by 4 because the output data type is int32. If not, the result might not be correct.